src/blockchain/alt_block_tree.cpp

```diff
@@ -173,76 +244,52 @@ void addContextToBlockIndex(BlockIndex<AltBlock>& index,
     }
   };
 
-  // add context from every VTB
-  for (const auto& vtb : p.vtbs) {
-    for (const auto& b : vtb.context) {
-      add(b);
-    }
-    add(vtb.containingBlock);
-  }
-
   // add context from ATV
   if (p.hasAtv) {
     for (const auto& b : p.atv.context) {
-      add(b);
+      addBlock(b);
+    }
+    addBlock(p.atv.containingBlock);
+  }
+
+  // step 2, process VTBs
+  for (const auto& vtb : p.vtbs) {
+    auto* temp = tree.getBlockIndex(vtb.getContainingBlock().getHash());
+
+    if (!temp || temp->containingEndorsements.find(
+                     BtcEndorsement::fromContainer(vtb).id) ==
+                     temp->containingEndorsements.end()) {
+      index.containingContext.top().vtbs.push_back(vtb);
     }
-    add(p.atv.containingBlock);
   }
 }
 
 template <>
 void removeContextFromBlockIndex(BlockIndex<AltBlock>& index,
                                  const BlockIndex<AltBlock>::payloads_t& p) {
-  auto& ctx = index.containingContext.vbk;
-  auto end = ctx.end();
-  auto remove = [&](const VbkBlock& b) {
-    end = std::remove(ctx.begin(), end, b);
+  if (index.containingContext.empty()) {
+    return;
+  }
+
+  auto& vbk = index.containingContext.top().vbk;
+  auto vbk_end = vbk.end();
+  auto removeBlock = [&](const VbkBlock& b) {
+    vbk_end = std::remove(vbk.begin(), vbk_end, b);
+  };
+
+  auto& vtbs = index.containingContext.top().vtbs;
+  auto vtbs_end = vtbs.end();
+  auto removeVTB = [&](const VTB& vtb) {
+    vtbs_end = std::remove(vtbs.begin(), vtbs_end, vtb);
   };
 
   // remove ATV context
-  std::for_each(p.atv.context.rbegin(), p.atv.context.rbegin(), remove);
+  std::for_each(p.atv.context.rbegin(), p.atv.context.rbegin(), removeBlock);
```

Comment:
Is it correct to have p.atv.context.rbegin() for both begin and end of the for_each?

Reply:
fixed